### PR TITLE
[CUDA] BiasSoftmax and Dropout Fusion

### PR DESCRIPTION
**Pull request by iK1D**
This PR is to fuse BiasSoftmax and Dropout and their gradients to BiasSoftmaxDropout and SoftmaxDropoutGrad for CUDA/ROCm EP. In detail:
- Vectorized and Non-vectorized BiasSoftmaxDropout and SoftmaxDropoutGrad CUDA/ROCm kernel
- Refactor of BiasSoftmax/SoftmaxGrad code, remove duplicated ROCm code files, use only one kernel launch function to launch kernels for different cases so that the new BiasSoftmaxDropout/SoftmaxDropoutGrad can call them directly.
- BiasSoftmaxDropoutFusion to fuse BiasSoftmax+Dropout to BiasSoftmaxDropout, DropoutGrad+BiasSoftmaxGrad to SoftmaxDropoutGrad.

Tested in A100 using one of the transformer models with 48 hidden layers:
Before, forward 260us + backward 192us = 452us
image
image

After, forward 216us + backward 130us = 346us, which is 1.3x faster.
image
image

It also contributes ~2% perf gain to the whole model.

### CODE REVIEW

**Comment by pengwa** [comment]
> 

Just make sure, from the PR description, I assume this new fusion will not affect our bitmask dropout replacement, is that right?

File: orttraining/orttraining/core/graph/training_op_defs.cc [hunk 1]
```diff
@@ -2192,6 +2192,65 @@ Example 4:
...
+      .Input(0, "data", "The input data as Tensor.", "T")
+      .Input(1, "bias", "The bias (or mask) as Tensor.", "T")
+      .Input(2, "ratio",
+             "The ratio of random dropout, with value in [0, 1). If this input was not set, "
+             "or if it was set to 0, the output would be a simple copy of the input. "
+             "If it's non-zero, output will be a random dropout of the scaled input, which is typically "
+             "the case during training. It is an optional value, if not specified it will default to 0.5.",
+             "T1", OpSchema::Optional)
+      .Output(0, "dropout_output", "The dropout output.", "T")
+      .Output(1, "mask", "The output mask of dropout.", "tensor(bool)")
```
Comment (pengwa) [inline orttraining/orttraining/core/graph/training_op_defs.cc:2219]: >"mask", "The  [](http://example.com/codeflow?start=17&length=13)

so we would lost the memory saving bitmask brought to us? is there a plan to make it a bit mask too?

File: orttraining/orttraining/core/graph/training_op_defs.cc [hunk 2]
```diff
@@ -2192,6 +2192,65 @@ Example 4:
...
+      .Input(1, "mask", "The mask output of the dropout.", "tensor(bool)")
+      .Input(2, "softmax_y", "The output of Softmax.", "T")
+      .Input(3, "ratio",
+             "Same value as the ratio input supplied to the dropout op with value in [0, 1). "
+             "If this input is not specified, a default value of 0.5 is used.",
+             "T1", OpSchema::Optional)
+      .Output(0, "dx", "Gradient of the input.", "T")
+      .TypeConstraint("T", {"tensor(float16)", "tensor(float)", "tensor(double)", "tensor(bfloat16)"},
+                      "Constrain input and output types to float tensors.")
+      .TypeConstraint("T1", {"tensor(float16)", "tensor(float)", "tensor(double)", "tensor(bfloat16)"},
```
Comment (pengwa) [inline orttraining/orttraining/core/graph/training_op_defs.cc:2251]: >TypeConstraint [](http://example.com/codeflow?start=7&length=14)

nit: shall we add shape infer logic for SoftmaxDropoutGrad as well.

File: orttraining/orttraining/core/optimizer/bias_softmax_dropout_fusion.cc [hunk 3]
```diff
@@ -0,0 +1,125 @@
...
+
+  for (auto node_index : node_topology_list) {
+    auto* node_ptr = graph.GetNode(node_index);
+    if (!node_ptr) continue;  // Node was removed.
+
+    auto& node = *node_ptr;
+    ORT_RETURN_IF_ERROR(Recurse(node, modified, graph_level, logger));
+
+    // Matching for BiasSoftmax node.
+    if (!graph_utils::IsSupportedOptypeVersionAndDomain(node, "BiasSoftmax", {1}, kMSDomain) ||
```
Comment (pengwa) [inline orttraining/orttraining/core/optimizer/bias_softmax_dropout_fusion.cc:27]: >BiasSoftmax [](http://example.com/codeflow?start=63&length=11)

softmax-11 + dropout can also be fused in the same kernel, I think you probably already had some thinking around that.

File: onnxruntime/core/providers/cuda/softmax.cc [hunk 4]
```diff
@@ -0,0 +1,28 @@
+// Copyright (c) Microsoft Corporation. All rights reserved.
+// Licensed under the MIT License.
+
+#include "core/providers/cuda/shared_inc/softmax.h"
+
+namespace onnxruntime {
```
Comment (pengwa) [inline onnxruntime/core/providers/cuda/softmax.cc:6]: >onnxruntime [](http://example.com/codeflow?start=10&length=11)

should we put this file in cuda/math?

File: onnxruntime/contrib_ops/cuda/math/bias_softmax_impl.cu [hunk 5]
```diff
@@ -215,7 +215,7 @@ Status BiasSoftmaxImpl(cudaStream_t stream, cudnnHandle_t cudnn_handle, T* outpu
                         static_cast<size_t>(batch_count * element_count));
 
   // invoke cuda DNN library for Y = softmax(X)
-  std::vector<int64_t> dims({batch_count, 1, 1, element_count});
+  const int64_t dims[]{batch_count, 1, 1, element_count};
```
Comment (pengwa) [inline onnxruntime/contrib_ops/cuda/math/bias_softmax_impl.cu:218]: >dims [](http://example.com/codeflow?start=16&length=4)

curious what is the intention for this change.

File: orttraining/orttraining/core/graph/training_op_defs.cc (hunk 1 above)
Comment (iK1D) [reply orttraining/orttraining/core/graph/training_op_defs.cc:2219]: > Just make sure, from the PR description, I assume this new fusion will not affect our bitmask dropout replacement, is that right?

It's not right. Fusing BiasSoftmaxDropout will replace BitmaskDropout to boolean Dropout. I don't figure out a way to support the BitmaskDropout in the softmax kernel when the element count for each batch cannot divided by 32, this means the some UNIT32 bitmask will cross different warps so that it's not easy to set the bitmask using warp APIs. But even switch back to boolean Dropout, the perf is still better than BiasSoftmax+BitmaskDropout (the perf comparison in the description is this case). So no plan for now to support bitmask as it's not an easy task.

File: orttraining/orttraining/core/graph/training_op_defs.cc (hunk 2 above)
Comment (iK1D) [reply orttraining/orttraining/core/graph/training_op_defs.cc:2251]: No need. Shape(GI) is same as Shape(I) (as well as data type) and our gradient builder will use the I's shape and type to set the GI's shape and type automatically for all gradient outputs.

File: orttraining/orttraining/core/optimizer/bias_softmax_dropout_fusion.cc (hunk 3 above)
Comment (iK1D) [reply orttraining/orttraining/core/optimizer/bias_softmax_dropout_fusion.cc:27]: I don't think it worth effort to support softmax-11 as its definition is out of date. It's possible we can fuse softmax-13 with axis=-1, but this requires extra checking in the fusion code, and also need a new Op SoftmaxDropout, or at least change the bias input optional and change the kernel code accordingly. This is doable but requires extra effort. But I think we can leave it as a model driven task, if we find such case in future model, I can add this support in the future.

File: onnxruntime/contrib_ops/cuda/math/bias_softmax_impl.cu (hunk 5 above)
Comment (iK1D) [reply onnxruntime/contrib_ops/cuda/math/bias_softmax_impl.cu:218]: Get rid of std::vector according to the coding standard in Dimitri's email.

File: orttraining/orttraining/core/graph/training_op_defs.cc (hunk 1 above)
Comment (pengwa) [reply orttraining/orttraining/core/graph/training_op_defs.cc:2219]: OK. This brings out the issue we talked previously; different fusion rules might bring different implications on compute latency or memory consumptions. And plus some recompute stuff, the best fusion strategy could be different on variant models. It's interesting to see such "conflict" in graph fusion.

File: orttraining/orttraining/core/graph/training_op_defs.cc (hunk 2 above)
Comment (pengwa) [reply orttraining/orttraining/core/graph/training_op_defs.cc:2251]: OK. I mean it from op schema perspective.

**Review by pengwa — APPROVED** [review]
:shipit:
